**Inclass material for Week 2: Exploratory Data Analysis**

This notebook was made based on main materials `2_Exploratory_Data_Analysis.ipynb`

Version: Newton - February 2021

___

# Exploratory Data Analysis (EDA)

## Training Objectives

The coursebook focuses on:
- Why and What: Exploratory Data Analysis
- Date Time objects
- Categorical data types
- Cross Tabulation and Pivot Table
- Treating Duplicates and Missing Values 

## Introduction to EDA

About 60 years ago, John Tukey defined data analysis as the "procedures for analyzing data, techniques for interpreting the results of such procedures ... and all the machinery of mathematical statistics which apply to analyzing data". His championing of EDA encouraged the development of statistical computing packages, especially S at Bell Labs (which later inspired R).

He wrote a book titled _Exploratory Data Analysis_ arguing that too much emphasis in statistics was placed on hypothesis testing (confirmatory data analysis) while not enough was placed on the discovery of the unexpected. 

> Exploratory data analysis isolates patterns and features of the data and reveals these forcefully to the analyst.

This course aims to present a selection of EDA techniques -- some developed by John Tukey himself -- but with a special emphasis on its application to modern business analytics.

## Data Loading

Don't forget to **import** required library before analysis:

In [77]:
import pandas as pd

Load `household` data from the directory `data_input/household.csv`:

In [78]:
household = pd.read_csv('data_input/household.csv')

In [79]:
household.head()

receipt_id  receipts_item_id    purchase_time category sub_category  \
0     9622257          32369294  7/22/2018 21:19     Rice         Rice   
1     9446359          31885876  7/15/2018 16:17     Rice         Rice   
2     9470290          31930241  7/15/2018 12:12     Rice         Rice   
3     9643416          32418582   7/24/2018 8:27     Rice         Rice   
4     9692093          32561236  7/26/2018 11:28     Rice         Rice   

        format  unit_price  discount  quantity yearmonth  
0  supermarket    128000.0         0         1   2018-07  
1   minimarket    102750.0         0         1   2018-07  
2  supermarket     64000.0         0         3   2018-07  
3   minimarket     65000.0         0         1   2018-07  
4  supermarket    124500.0         0         1   2018-07

Data terkait transaksi Rice, Detergent (Fabric Care), dan Sugar 

In [80]:
household['category'].unique()

array(['Rice', 'Fabric Care', 'Sugar/Flavored Syrup'], dtype=object)

💡 **Tips**: Using `.info()` method, we can check complete **information** of our DataFrame:

- Number of rows and columns (`.shape`)
- Column name and number of non-null value (`.columns`)
- Data types of each column (`.dtypes`)
- Memory usage

In [81]:
household.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 72000 entries, 0 to 71999
Data columns (total 10 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   receipt_id        72000 non-null  int64  
 1   receipts_item_id  72000 non-null  int64  
 2   purchase_time     72000 non-null  object 
 3   category          72000 non-null  object 
 4   sub_category      72000 non-null  object 
 5   format            72000 non-null  object 
 6   unit_price        72000 non-null  float64
 7   discount          72000 non-null  int64  
 8   quantity          72000 non-null  int64  
 9   yearmonth         72000 non-null  object 
dtypes: float64(1), int64(4), object(5)
memory usage: 5.5+ MB


Note: 72000 non-null untuk semua kolom menandakan data `household` tidak ada missing value

___

# Working with Datetime

Let's focus on column data types of `household`:

In [82]:
household.dtypes

receipt_id            int64
receipts_item_id      int64
purchase_time        object
category             object
sub_category         object
format               object
unit_price          float64
discount              int64
quantity              int64
yearmonth            object
dtype: object

Which column should be stored in `datetime64` data type?

> Answer: `purchase_time`

In [83]:
# apakah yearmonth dijadikan datetime64?
household['yearmonth']
# jawaban: category, karena tidak ada komponen day (hanya tahun dan bulan), dan juga nilainya berulang
# datetime64 -> yyyy-mm-dd

0        2018-07
1        2018-07
2        2018-07
3        2018-07
4        2018-07
          ...   
71995    2017-12
71996    2017-12
71997    2017-12
71998    2017-12
71999    2017-12
Name: yearmonth, Length: 72000, dtype: object

## Convert to Datetime

There are three common ways for us to convert data type into `datetime64`:

### First: Method `.astype()`

Let's create a copy of `household` so that the original data remains unchanged.

In [84]:
df_1 = household.copy()
df_1.dtypes

receipt_id            int64
receipts_item_id      int64
purchase_time        object
category             object
sub_category         object
format               object
unit_price          float64
discount              int64
quantity              int64
yearmonth            object
dtype: object

Convert data type using `.astype()` method:

In [85]:
df_1['purchase_time'] = df_1['purchase_time'].astype('datetime64')

In [86]:
df_1.dtypes

receipt_id                   int64
receipts_item_id             int64
purchase_time       datetime64[ns]
category                    object
sub_category                object
format                      object
unit_price                 float64
discount                     int64
quantity                     int64
yearmonth                   object
dtype: object

⚠️ **Warning**: Don't forget to assign the result to its original column.

### Second: Parameter `parse_dates`

Used during data loading, assuming we already know which columns are supposed to be `datetime64`.

In [87]:
df_2 = pd.read_csv('data_input/household.csv', parse_dates=['purchase_time'])
df_2.dtypes

receipt_id                   int64
receipts_item_id             int64
purchase_time       datetime64[ns]
category                    object
sub_category                object
format                      object
unit_price                 float64
discount                     int64
quantity                     int64
yearmonth                   object
dtype: object

In [88]:
df_2['purchase_time'].head()
# format datetime64 adalah yyyy-mm-dd hh:mm:ss

0   2018-07-22 21:19:00
1   2018-07-15 16:17:00
2   2018-07-15 12:12:00
3   2018-07-24 08:27:00
4   2018-07-26 11:28:00
Name: purchase_time, dtype: datetime64[ns]

### Third: Method `pd.to_datetime()`

In [89]:
df_3 = household.copy()
df_3.dtypes

receipt_id            int64
receipts_item_id      int64
purchase_time        object
category             object
sub_category         object
format               object
unit_price          float64
discount              int64
quantity              int64
yearmonth            object
dtype: object

Convert data type using `pd.to_datetime()` method:

In [90]:
df_3['purchase_time'] = pd.to_datetime(df_3['purchase_time'])

In [91]:
df_3.dtypes

receipt_id                   int64
receipts_item_id             int64
purchase_time       datetime64[ns]
category                    object
sub_category                object
format                      object
unit_price                 float64
discount                     int64
quantity                     int64
yearmonth                   object
dtype: object

So, what is the difference between using `.astype()` and `pd.to_datetime()`?

Unlike using `astype()`, with `pd.to_datetime()` we are allowed to specify **parameters** for the datetime conversion. Thus, provide more **flexibility**.

Suppose we have a column which stores a daily sales data from end of January to the beginning of February:

In [92]:
sales_date = pd.Series(['30-01-2021', '31-01-2021', '01-02-2021', '02-02-2021'])
sales_date

0    30-01-2021
1    31-01-2021
2    01-02-2021
3    02-02-2021
dtype: object

The example above shows how Indonesians usually write dates, using **dd-mm-yyyy** format. Let's see what will happen when we convert `sales_date` data type to `datetime64`:

In [93]:
sales_date.astype('datetime64') # yyyy-mm-dd

0   2021-01-30
1   2021-01-31
2   2021-01-02
3   2021-02-02
dtype: datetime64[ns]

Take a look on the third observation (second index):

- Expectation: 1 February
- Reality: 2 January

⚠️ **Warning**: By default, `pandas` will infer date as **month first** format.

#### Parameter `dayfirst`

Solution 1: Use parameter `dayfirst=True` to specify that our `sales_date` starts with day, not month.

In [94]:
pd.to_datetime(sales_date, dayfirst=True)

0   2021-01-30
1   2021-01-31
2   2021-02-01
3   2021-02-02
dtype: datetime64[ns]

#### Parameter `format`

Solution 2: Use parameter `format` to specifically specify the date formatting of our `sales_date`.

In [95]:
pd.to_datetime(sales_date, format='%d-%m-%Y')

0   2021-01-30
1   2021-01-31
2   2021-02-01
3   2021-02-02
dtype: datetime64[ns]

- `%d`: tanggal, dua digit
- `%m`: bulan, dua digit
- `%Y`: tahun, empat digit

The specified `format` is called as Python's `strftime` directives.

`strftime` means string from time, we can format the time in different desirable ways.

[Documentation of `strftime`](https://strftime.org/)

Misalkan format tanggal yang kita miliki sekarang `dd/mm-yyyy` maka formatnya juga mengikuti: `format='%d/%m-%Y'`

In [96]:
sales_date2 = pd.Series(['30/01-2021', '31/01-2021', '01/02-2021', '02/02-2021'])
pd.to_datetime(sales_date2, format='%d/%m-%Y')

0   2021-01-30
1   2021-01-31
2   2021-02-01
3   2021-02-02
dtype: datetime64[ns]

Note:

- Untuk `datetime64` formatnya selalu yyyy-mm-dd hh:mm:ss
- `ns`: nanosecond

Misalkan format tanggal dan bulan tanpa '0' melainkan langsung angkanya, maka formatnya sama saja:

In [97]:
sales_date3 = pd.Series(['1-1-2021', '2-1-2021', '3-1-2021', '4-1-2021'])
pd.to_datetime(sales_date3, format='%d-%m-%Y')

0   2021-01-01
1   2021-01-02
2   2021-01-03
3   2021-01-04
dtype: datetime64[ns]

## Extract Datetime Component

The main reason we convert a column to `datetime64` date type because `pandas` has a number of machineries to work with `datetime` objects. These are convenient for when we need to extract datetime component.

**Date component (numeric)**
- `.dt.year`
- `.dt.month`
- `.dt.day`
- `.dt.dayofweek`: index of day, Monday = 0 to Sunday = 6

**Date component (string)**
- `.dt.month_name()`
- `.dt.day_name()`

**Time component**
- `.dt.hour`
- `.dt.minute`
- `.dt.second`

[Documentation of datetime properties](https://pandas.pydata.org/pandas-docs/stable/reference/series.html#datetimelike-properties)

Let's extract datetime component from `household`, but first let's convert `purchase_time` column to `datetime64`:

In [98]:
# your code here
household['purchase_time'] = household['purchase_time'].astype('datetime64')
household['purchase_time']

0       2018-07-22 21:19:00
1       2018-07-15 16:17:00
2       2018-07-15 12:12:00
3       2018-07-24 08:27:00
4       2018-07-26 11:28:00
                ...        
71995   2017-12-27 09:20:00
71996   2017-12-13 19:52:00
71997   2017-12-27 08:03:00
71998   2017-12-07 12:29:00
71999   2017-12-19 18:59:00
Name: purchase_time, Length: 72000, dtype: datetime64[ns]

In [99]:
household.dtypes

receipt_id                   int64
receipts_item_id             int64
purchase_time       datetime64[ns]
category                    object
sub_category                object
format                      object
unit_price                 float64
discount                     int64
quantity                     int64
yearmonth                   object
dtype: object

Check the range of `purchase_time`:

In [100]:
household['purchase_time'].describe()

<ipython-input-100-ae4a8afa6a47>:1: FutureWarning: Treating datetime data as categorical rather than numeric in `.describe` is deprecated and will be removed in a future version of pandas. Specify `datetime_is_numeric=True` to silence this warning and adopt the future behavior now.
  household['purchase_time'].describe()


count                   72000
unique                  62072
top       2017-10-22 12:00:00
freq                       12
first     2017-10-01 00:00:00
last      2018-09-30 23:57:00
Name: purchase_time, dtype: object

Note: `dtype: object` bukanlah tipe data dari kolom `purchase_time`, melainkan dari hasil `.describe()`

### Datetime attribute

To extract datetime component in numeric value, we can use its **attribute** (without bracket)

In [101]:
household['purchase_time']

0       2018-07-22 21:19:00
1       2018-07-15 16:17:00
2       2018-07-15 12:12:00
3       2018-07-24 08:27:00
4       2018-07-26 11:28:00
                ...        
71995   2017-12-27 09:20:00
71996   2017-12-13 19:52:00
71997   2017-12-27 08:03:00
71998   2017-12-07 12:29:00
71999   2017-12-19 18:59:00
Name: purchase_time, Length: 72000, dtype: datetime64[ns]

In [102]:
household['purchase_time'].dt.month

0         7
1         7
2         7
3         7
4         7
         ..
71995    12
71996    12
71997    12
71998    12
71999    12
Name: purchase_time, Length: 72000, dtype: int64

In [103]:
household['purchase_time'].dt.hour

0        21
1        16
2        12
3         8
4        11
         ..
71995     9
71996    19
71997     8
71998    12
71999    18
Name: purchase_time, Length: 72000, dtype: int64

### Datetime method

To extract name (text) from datetime component, we can use its **method** (with bracket)

In [104]:
household['purchase_time'].dt.day_name()

0           Sunday
1           Sunday
2           Sunday
3          Tuesday
4         Thursday
           ...    
71995    Wednesday
71996    Wednesday
71997    Wednesday
71998     Thursday
71999      Tuesday
Name: purchase_time, Length: 72000, dtype: object

In [105]:
household['purchase_time'].dt.month_name()

0            July
1            July
2            July
3            July
4            July
           ...   
71995    December
71996    December
71997    December
71998    December
71999    December
Name: purchase_time, Length: 72000, dtype: object

## Datetime Transformation

Supposed we want to transform the existing `datetime64` column into values of periods, we can use the `.to_period()` method:

- `.dt.to_period('D')`: Daily (yyyy-mm-dd)
- `.dt.to_period('W')`: Weekly (Monday to Sunday period)
- `.dt.to_period('M')`: Monthly (year-month)
- `.dt.to_period('Q')`: Quarterly (year-quarter)

[Documentation of offset aliases](https://pandas.pydata.org/pandas-docs/stable/user_guide/timeseries.html#timeseries-offset-aliases)

In [106]:
household['purchase_time'].dt.to_period('D') # Daily

0        2018-07-22
1        2018-07-15
2        2018-07-15
3        2018-07-24
4        2018-07-26
            ...    
71995    2017-12-27
71996    2017-12-13
71997    2017-12-27
71998    2017-12-07
71999    2017-12-19
Name: purchase_time, Length: 72000, dtype: period[D]

In [107]:
household['purchase_time'].dt.to_period('W') # Weekly: Monday/Sunday

0        2018-07-16/2018-07-22
1        2018-07-09/2018-07-15
2        2018-07-09/2018-07-15
3        2018-07-23/2018-07-29
4        2018-07-23/2018-07-29
                 ...          
71995    2017-12-25/2017-12-31
71996    2017-12-11/2017-12-17
71997    2017-12-25/2017-12-31
71998    2017-12-04/2017-12-10
71999    2017-12-18/2017-12-24
Name: purchase_time, Length: 72000, dtype: period[W-SUN]

In [108]:
household['purchase_time'].dt.to_period('W-SAT') # Weekly Saturday: Sunday/Saturday

0        2018-07-22/2018-07-28
1        2018-07-15/2018-07-21
2        2018-07-15/2018-07-21
3        2018-07-22/2018-07-28
4        2018-07-22/2018-07-28
                 ...          
71995    2017-12-24/2017-12-30
71996    2017-12-10/2017-12-16
71997    2017-12-24/2017-12-30
71998    2017-12-03/2017-12-09
71999    2017-12-17/2017-12-23
Name: purchase_time, Length: 72000, dtype: period[W-SAT]

In [109]:
household['purchase_time'].dt.to_period('M') # Monthly: year-month

0        2018-07
1        2018-07
2        2018-07
3        2018-07
4        2018-07
          ...   
71995    2017-12
71996    2017-12
71997    2017-12
71998    2017-12
71999    2017-12
Name: purchase_time, Length: 72000, dtype: period[M]

In [110]:
household['purchase_time'].dt.to_period('Q') # Quarterly: yearQ[1/2/3/4]

0        2018Q3
1        2018Q3
2        2018Q3
3        2018Q3
4        2018Q3
          ...  
71995    2017Q4
71996    2017Q4
71997    2017Q4
71998    2017Q4
71999    2017Q4
Name: purchase_time, Length: 72000, dtype: period[Q-DEC]

**In class Question 1**: Bagaimana untuk menggabungkan beberapa kolom komponen tanggal menjadi satu?

Misal pakai data `companies.csv` kita gabungkan kolom `Month`, `Day`, dan `Year`:

In [111]:
companies = pd.read_csv("./data_input/companies.csv")
companies

ID                 Customer Name Consulting Sales Software Sales  \
0  30940               New Media Group       IDR7125000     IDR5500000   
1  82391               Li and Partners        IDR420000      IDR820000   
2  18374      PT. Kreasi Metrik Solusi                0      IDR550403   
3  57531  PT. Algoritma Data Indonesia        IDR850000      IDR395500   
4  19002         Palembang Konsultansi       IDR2115000              0   
5  31142     PT. Surya Citra Manajemen        IDR960000      IDR503000   

  Forecasted Growth       Returns  Month  Day  Year  Location     Account  
0            30.00%  IDR1,500,000      1   10  2017   Jakarta  Enterprise  
1            10.00%    IDR400,000      6   15  2016   Jakarta     Startup  
2            25.00%             0      3   29  2012  Surabaya  Enterprise  
3             4.00%             0      7   17  2017   Jakarta     Startup  
4           -15.00%             0      2   24  2018   Bandung     Startup  
5            19.00%             0      1   19  2019   Jakarta  Enterprise

Tahapan:

1. Gabungkan dengan operasi `+`, namun pastikan tipenya **str** (string) agar dapat digabungkan
2. Ubah menjadi tipe data `datetime64`
3. Kemudian dapat diekstrak komponen datetime nya

In [112]:
companies['date'] = companies['Month'].astype(str) + '-' + companies['Day'].astype(str) + '-' + companies['Year'].astype(str)
companies['date'] = companies['date'].astype('datetime64')
companies['date'].dt.day_name()

0      Tuesday
1    Wednesday
2     Thursday
3       Monday
4     Saturday
5     Saturday
Name: date, dtype: object

**In class Question 2**: Apakah str = object?

Tipe data object umumnya adalah string (kalimat) pada file text/csv. Object tidak selalu string, bahkan kita dapat meletakkan DataFrame di dalam DataFrame, dan DataFrame yang ada di dalam dikenal sebagai object. Namun hal ini jarang dilakukan dalam melakukan analisis data.

In [113]:
tabel1 = pd.DataFrame({'kolom1': [1, 2, 3],
                       'kolom2': ['a', 'b', 'c']})
tabel2 = tabel1.copy()

tabel = pd.DataFrame({'kolom1': [tabel1, tabel2]})
tabel

kolom1
0     kolom1 kolom2
0       1      a
1       2   ...
1     kolom1 kolom2
0       1      a
1       2   ...

In [114]:
tabel.dtypes

kolom1    object
dtype: object

## Knowledge Check: Datetime

_Estimated time required: 15 minutes_

1. In the following cell, start again by reading in the `household.csv` dataset.
2. Make sure the `purchase_time` column has converted as a datetime object.
3. Use `x.dt.day_name()`, assuming `x` is a datetime object to get the day of week. Assign this to a new column in your `household` Data Frame, name it `weekday`
4. The `yearmonth` column stores the information of year and month of the `purchase_time`. Using `dt.to_period()`, how will you recreate the column if you needed the same information?
5. Print the first 5 rows of your data to verify that your preprocessing steps are correct

Alternatif untuk nomor 2:

- Menggunakan `household['purchase_time'].astype('datetime64')`
- Menambahkan parameter `parse_dates = ['purchase_time']` di dalam `pd.read_csv()`

In [115]:
# your code here
household = pd.read_csv('data_input/household.csv')
household['purchase_time'] = pd.to_datetime(household['purchase_time'])
household['weekday'] = household['purchase_time'].dt.day_name()
household['yearmonth_recreate'] = household['purchase_time'].dt.to_period('M') # Monthly: year-month
household.head()

receipt_id  receipts_item_id       purchase_time category sub_category  \
0     9622257          32369294 2018-07-22 21:19:00     Rice         Rice   
1     9446359          31885876 2018-07-15 16:17:00     Rice         Rice   
2     9470290          31930241 2018-07-15 12:12:00     Rice         Rice   
3     9643416          32418582 2018-07-24 08:27:00     Rice         Rice   
4     9692093          32561236 2018-07-26 11:28:00     Rice         Rice   

        format  unit_price  discount  quantity yearmonth   weekday  \
0  supermarket    128000.0         0         1   2018-07    Sunday   
1   minimarket    102750.0         0         1   2018-07    Sunday   
2  supermarket     64000.0         0         3   2018-07    Sunday   
3   minimarket     65000.0         0         1   2018-07   Tuesday   
4  supermarket    124500.0         0         1   2018-07  Thursday   

  yearmonth_recreate  
0            2018-07  
1            2018-07  
2            2018-07  
3            2018-07  
4            2018-07

💭 **Bonus challenge:**

Suppose that the estimated delivery time will take exactly 2 days after the product is purchased. How do we obtain the result by using `purchase_time` column?

In [116]:
household['purchase_time']

0       2018-07-22 21:19:00
1       2018-07-15 16:17:00
2       2018-07-15 12:12:00
3       2018-07-24 08:27:00
4       2018-07-26 11:28:00
                ...        
71995   2017-12-27 09:20:00
71996   2017-12-13 19:52:00
71997   2017-12-27 08:03:00
71998   2017-12-07 12:29:00
71999   2017-12-19 18:59:00
Name: purchase_time, Length: 72000, dtype: datetime64[ns]

In [117]:
# Pak Prasetyo, Pak Zainul, Bu Rahmah
from datetime import timedelta
household['purchase_time'] + timedelta(days=2)

0       2018-07-24 21:19:00
1       2018-07-17 16:17:00
2       2018-07-17 12:12:00
3       2018-07-26 08:27:00
4       2018-07-28 11:28:00
                ...        
71995   2017-12-29 09:20:00
71996   2017-12-15 19:52:00
71997   2017-12-29 08:03:00
71998   2017-12-09 12:29:00
71999   2017-12-21 18:59:00
Name: purchase_time, Length: 72000, dtype: datetime64[ns]

In [118]:
# Pak Privera
household['purchase_time'] + pd.Timedelta(2, 'D') # 2 Days

0       2018-07-24 21:19:00
1       2018-07-17 16:17:00
2       2018-07-17 12:12:00
3       2018-07-26 08:27:00
4       2018-07-28 11:28:00
                ...        
71995   2017-12-29 09:20:00
71996   2017-12-15 19:52:00
71997   2017-12-29 08:03:00
71998   2017-12-09 12:29:00
71999   2017-12-21 18:59:00
Name: purchase_time, Length: 72000, dtype: datetime64[ns]

In [119]:
# Pak Endi: bisa, tetapi untuk data timenya hilang
household['purchase_time'].dt.to_period('D') + 2

0        2018-07-24
1        2018-07-17
2        2018-07-17
3        2018-07-26
4        2018-07-28
            ...    
71995    2017-12-29
71996    2017-12-15
71997    2017-12-29
71998    2017-12-09
71999    2017-12-21
Name: purchase_time, Length: 72000, dtype: period[D]

## Timedelta Object

`pandas` has `.Timedelta()` method which represents a duration, the difference between two dates or times.

[Documentation of Timedelta](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Timedelta.html#pandas-timedelta)

In [120]:
# contoh selisih antar dua tanggal
contoh = pd.DataFrame({
    'id': [1, 2, 3],
    'start': ['2021-01-01 01:00:00', '2021-01-02 01:00:00', '2021-01-03 01:00:00'],
    'end': ['2021/01/04 15:00:00', '2021/01/07 16:00:00', '2021/01/08 17:00:00']
})

contoh[['start', 'end']] = contoh[['start', 'end']].astype('datetime64')
contoh['end'] - contoh['start']

0   3 days 14:00:00
1   5 days 15:00:00
2   5 days 16:00:00
dtype: timedelta64[ns]

In [121]:
(contoh['end'] - contoh['start']).dt.days

0    3
1    5
2    5
dtype: int64

**END OF DAY 1**

___

**START OF DAY 2**

# Working with Category

## Characteristic: Unique Values

Characteristic of `category` data type: repeated values, which can be categorized into several groups. We can use:

- `.unique()` to see unique values of a Series
- `.nunique()` to see number of unique values of a Series or DataFrame

to identify which columns are better converted to `category`.

In [122]:
household.nunique()

receipt_id            69776
receipts_item_id      72000
purchase_time         62072
category                  3
sub_category              3
format                    3
unit_price             3884
discount               1329
quantity                 19
yearmonth                12
weekday                   7
yearmonth_recreate       12
dtype: int64

In [123]:
cat_col = ['category', 'sub_category', 'format', 'yearmonth']

household[cat_col] = household[cat_col].astype('category')

## Convert to Category

Let's create a copy of `household` so that the original data remains unchanged.

In [124]:
household_cat = household.copy()
household_cat.dtypes

receipt_id                     int64
receipts_item_id               int64
purchase_time         datetime64[ns]
category                    category
sub_category                category
format                      category
unit_price                   float64
discount                       int64
quantity                       int64
yearmonth                   category
weekday                       object
yearmonth_recreate         period[M]
dtype: object

Let's convert the columns to `category` data type:

## Advantages

There are two main advantages of converting to `category`:

### First: Memory Efficient

We can compare two Data Frame **before and after** the columns are converted to `category` data type:

- `household` (before): 6.0+ MB
- `household_cat` (after): 3.6 MB

In [125]:
household.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 72000 entries, 0 to 71999
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   receipt_id          72000 non-null  int64         
 1   receipts_item_id    72000 non-null  int64         
 2   purchase_time       72000 non-null  datetime64[ns]
 3   category            72000 non-null  category      
 4   sub_category        72000 non-null  category      
 5   format              72000 non-null  category      
 6   unit_price          72000 non-null  float64       
 7   discount            72000 non-null  int64         
 8   quantity            72000 non-null  int64         
 9   yearmonth           72000 non-null  category      
 10  weekday             72000 non-null  object        
 11  yearmonth_recreate  72000 non-null  period[M]     
dtypes: category(4), datetime64[ns](1), float64(1), int64(4), object(1), period[M](1)
memory usage: 4.7+ MB

In [126]:
household_cat.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 72000 entries, 0 to 71999
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   receipt_id          72000 non-null  int64         
 1   receipts_item_id    72000 non-null  int64         
 2   purchase_time       72000 non-null  datetime64[ns]
 3   category            72000 non-null  category      
 4   sub_category        72000 non-null  category      
 5   format              72000 non-null  category      
 6   unit_price          72000 non-null  float64       
 7   discount            72000 non-null  int64         
 8   quantity            72000 non-null  int64         
 9   yearmonth           72000 non-null  category      
 10  weekday             72000 non-null  object        
 11  yearmonth_recreate  72000 non-null  period[M]     
dtypes: category(4), datetime64[ns](1), float64(1), int64(4), object(1), period[M](1)
memory usage: 4.7+ MB

### Second: Categorical Accessor `.cat`

Just like the `datetime64` data type which has `.dt` accessor, the `category` data type has `.cat` accessor.

In [127]:
household_cat['format'].cat.categories

Index(['hypermarket', 'minimarket', 'supermarket'], dtype='object')

You can explore more functionalities by refering to [documentation of categorical accessor](https://pandas.pydata.org/pandas-docs/stable/reference/series.html#categorical-accessor) for the complete list.

___

# Contingency Table

One of the simplest EDA toolkit to display **counts** of a categorical column:

Contingency = Cross-tabulation = Frequency tables

In [151]:
household = household_cat.copy()
household.dtypes

receipt_id                     int64
receipts_item_id               int64
purchase_time         datetime64[ns]
category                    category
sub_category                category
format                      category
unit_price                   float64
discount                       int64
quantity                       int64
yearmonth                   category
weekday                       object
yearmonth_recreate         period[M]
dtype: object

## Method `.value_counts()`

Usage: Get the counts of each unique levels in one categorical column, sorted by descending order.

Parameter:

- `sort=False`: prevent any sorting values, **sort by index** instead
- `ascending=True`: **sort values** in ascending order

<b id='q1'>Business Question 1</b>

There are three categories of `format` (market type), how many total transactions occurred in each `format`?

In [129]:
household['format'].unique()

# count_unique = len(pd.unique(household['format'])) 
# count_unique

['supermarket', 'minimarket', 'hypermarket']
Categories (3, object): ['supermarket', 'minimarket', 'hypermarket']

In [130]:
household['format'].value_counts()

minimarket     46803
supermarket    19826
hypermarket     5371
Name: format, dtype: int64

Let's say we don't need sort by values:

In [131]:
household.head

<bound method NDFrame.head of        receipt_id  receipts_item_id       purchase_time              category  \
0         9622257          32369294 2018-07-22 21:19:00                  Rice   
1         9446359          31885876 2018-07-15 16:17:00                  Rice   
2         9470290          31930241 2018-07-15 12:12:00                  Rice   
3         9643416          32418582 2018-07-24 08:27:00                  Rice   
4         9692093          32561236 2018-07-26 11:28:00                  Rice   
...           ...               ...                 ...                   ...   
71995     5909305          17998610 2017-12-27 09:20:00  Sugar/Flavored Syrup   
71996     5736299          17432379 2017-12-13 19:52:00  Sugar/Flavored Syrup   
71997     5901144          18263665 2017-12-27 08:03:00  Sugar/Flavored Syrup   
71998     5660630          17222218 2017-12-07 12:29:00  Sugar/Flavored Syrup   
71999     5808147          17658789 2017-12-19 18:59:00  Sugar/Flavored Syrup  

Sort by ascending (smallest to largest):

In [301]:
household['weekday'].head()

0      Sunday
1      Sunday
2      Sunday
3     Tuesday
4    Thursday
Name: weekday, dtype: category
Categories (7, object): ['Monday' < 'Tuesday' < 'Wednesday' < 'Thursday' < 'Friday' < 'Saturday' < 'Sunday']

<b id='q2'>Business Question 2</b>

Which `weekday` has the **largest** and **smallest** transaction volume?

In [300]:
household['weekday'].value_counts().max()

Sunday       12573
Saturday     11828
Friday       10778
Tuesday       9427
Wednesday     9206
Thursday      9138
Monday        9050
Name: weekday, dtype: int64

In [149]:
household['weekday'].value_counts().min()

9050

## Cross-Tabulation

Versatile solution in producing frequency table is by using `crosstab`. Syntax:

    pd.crosstab(
        index=...,
        columns=...
    )
                
**Required Parameter**:

- `index`: Values to group by in the index (rows)
- `columns`: Values to group by in the columns

Let's re-create [Business Question 1](#q1) using `pd.crosstab()`:

How many total transactions occurred in each `format`?

In [164]:
crosstab = pd.crosstab(
    index=household['format'],
    columns='Total Transactions')

crosstab

col_0        Total Transactions
format                         
hypermarket                5371
minimarket                46803
supermarket               19826

Let's re-create [Business Question 2](#q2) using `pd.crosstab()`:

Which `weekday` has the **largest** and **smallest** transaction volume?

In [171]:
weekday_summary = pd.crosstab(
    index=household['weekday'],
    columns='Most Popular Week')

weekday_summary

col_0      Most Popular Week
weekday                     
Friday                 10778
Monday                  9050
Saturday               11828
Sunday                 12573
Thursday                9138
Tuesday                 9427
Wednesday               9206

To get the insight easier, we have to sort the DataFrame by using `.sort_values()` method. Let's sort the table above in **descending** order:

In [191]:
household['weekday'] = household['weekday'].astype('category')

Optional: If you want to sort by `weekday` name, using `.cat` accessor:

In [189]:
household['weekday'] = household['weekday'].cat.reorder_categories(
    ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday'],
    ordered=True
)

In [188]:
household['weekday']

0           Sunday
1           Sunday
2           Sunday
3          Tuesday
4         Thursday
           ...    
71995    Wednesday
71996    Wednesday
71997    Wednesday
71998     Thursday
71999      Tuesday
Name: weekday, Length: 72000, dtype: category
Categories (7, object): ['Monday' < 'Tuesday' < 'Wednesday' < 'Thursday' < 'Friday' < 'Saturday' < 'Sunday']

<b id='q3'>Business Question 3</b>

From [Business Question 1](#q1) we know that minimarket has the most frequent transaction. Let's say we are curious to know what `category` of items has the highest total transactions in minimarket?  

In [197]:
category_summary = pd.crosstab(
    index=household['category'],
    columns='Most Popular Category').sort_values(by='Most Popular Category', ascending=False).head(1)

category_summary

col_0        Most Popular Category
category                          
Fabric Care                  36000

In [204]:
category_summary = pd.crosstab(
    index=household['category'],
    columns=household['format']).sort_values('minimarket', ascending=False)

category_summary

format                hypermarket  minimarket  supermarket
category                                                  
Fabric Care                  2611       24345         9044
Sugar/Flavored Syrup         1761       15370         6869
Rice                          999        7088         3913

In [202]:
minimarket = household[household['format'] == 'minimarket']

pd.crosstab(
    index=minimarket['category'],
    columns='Total transaction in minimarket'
)

col_0                 Total transaction in minimarket
category                                             
Fabric Care                                     24345
Rice                                             7088
Sugar/Flavored Syrup                            15370

### Additional Parameters of `pd.crosstab()`

#### Margins

Usage: Calculate subtotals

Parameter

- `margins`: Add row and column margins for subtotals
- `margins_name`: Name of the row and column that will contain the totals when `margins=True`

In [207]:
pd.crosstab(
    index=household['category'],
    columns=household['format'],
    margins=True 
)

format                hypermarket  minimarket  supermarket    All
category                                                         
Fabric Care                  2611       24345         9044  36000
Rice                          999        7088         3913  12000
Sugar/Flavored Syrup         1761       15370         6869  24000
All                          5371       46803        19826  72000

In [208]:
pd.crosstab(
    index=household['format'],
    columns=household['category'],
    margins=True 
)

category     Fabric Care   Rice  Sugar/Flavored Syrup    All
format                                                      
hypermarket         2611    999                  1761   5371
minimarket         24345   7088                 15370  46803
supermarket         9044   3913                  6869  19826
All                36000  12000                 24000  72000

In [139]:
# change margins_name
pd.crosstab(
    index=household['category'],
    columns=household['format'],
    margins=True,
    margins_name='Total Transaction'
)

format                hypermarket  minimarket  supermarket  Total Transaction
category                                                                     
Fabric Care                  2611       24345         9044              36000
Rice                          999        7088         3913              12000
Sugar/Flavored Syrup         1761       15370         6869              24000
Total Transaction            5371       46803        19826              72000

#### Normalize

Usage: Calculate percentages/proportions instead of the frequency

- `normalize='all'` or `normalize=True`: normalize over all values
- `normalize='index'`: normalize over each row
- `normalize='columns'`: normalize over each column

**Normalize by all**

In [212]:
pd.crosstab(
    index=household['category'],
    columns=household['format'],
    margins=True,
    normalize='all'
)

format                hypermarket  minimarket  supermarket       All
category                                                            
Fabric Care              0.036264    0.338125     0.125611  0.500000
Rice                     0.013875    0.098444     0.054347  0.166667
Sugar/Flavored Syrup     0.024458    0.213472     0.095403  0.333333
All                      0.074597    0.650042     0.275361  1.000000

**Normalize by index**

In [213]:
pd.crosstab(
    index=household['category'],
    columns=household['format'],
    margins=True,
    normalize='index'
)

format                hypermarket  minimarket  supermarket
category                                                  
Fabric Care              0.072528    0.676250     0.251222
Rice                     0.083250    0.590667     0.326083
Sugar/Flavored Syrup     0.073375    0.640417     0.286208
All                      0.074597    0.650042     0.275361

**Normalize by column**

In [214]:
pd.crosstab(
    index=household['category'],
    columns=household['format'],
    margins=True,
    normalize='columns'
)

format                hypermarket  minimarket  supermarket       All
category                                                            
Fabric Care              0.486129    0.520159     0.456169  0.500000
Rice                     0.185999    0.151443     0.197367  0.166667
Sugar/Flavored Syrup     0.327872    0.328398     0.346464  0.333333

## Knowledge Check: Contingency Table

_Estimated time required: 15 minutes_

1. In which period (`yearmonth`) does the hypermarket achieve the highest total transactions?

In [395]:
# your code here
# category_summary = pd.crosstab(
#     index=household['category'],
#     columns=household['format']).sort_values('minimarket', ascending=False)

# category_summary

pd.crosstab(
    index=household['yearmonth'],
    columns=household['format']
).sort_values('hypermarket', ascending=False).head(3)

format     hypermarket  minimarket  supermarket
yearmonth                                      
2018-03            521        3540         1939
2018-02            513        3745         1742
2018-05            487        3704         1809

2. In the second quarter of 2018, how many percentage of total transactions came from the supermarket, when we compared to other `format`? (Hint: extract the quarterly period from `purchase_time`)

In [401]:
# your code here

household['quarter'] = household['purchase_time'].dt.to_period('Q')

x = pd.crosstab(
    index=household['quarter'],
    columns=household['format'],
    normalize='index'
).sort_values('supermarket', ascending=False)*100

x.loc[['2018Q2'], ['supermarket']]

format   supermarket
quarter             
2018Q2     28.427778

💭 **Bonus challenge:**

Produce a frequency table where we can breakdown the total transactions for each `yearmonth` period by each product `category` in each market (`format`).

In [402]:
pd.crosstab(
    index=household['yearmonth'],
    columns=[household['category'], household['format']],
)

category  Fabric Care                               Rice             \
format    hypermarket minimarket supermarket hypermarket minimarket   
yearmonth                                                             
2017-10           213       2009         778          81        600   
2017-11           218       2069         713          76        581   
2017-12           206       2111         683          68        583   
2018-01           229       2094         677          68        597   
2018-02           212       1989         799         117        549   
2018-03           249       1853         898          97        559   
2018-04           176       2031         793          79        578   
2018-05           236       1884         880          82        585   
2018-06           230       1996         774          81        622   
2018-07           169       2289         542          65        626   
2018-08           253       2025         722          82        611   
2018-09           220       1995         785         103        597   

category              Sugar/Flavored Syrup                         
format    supermarket          hypermarket minimarket supermarket  
yearmonth                                                          
2017-10           319                  152       1289         559  
2017-11           343                  148       1262         590  
2017-12           349                  153       1245         602  
2018-01           335                  152       1269         579  
2018-02           334                  184       1207         609  
2018-03           344                  175       1128         697  
2018-04           343                  121       1300         579  
2018-05           333                  169       1235         596  
2018-06           297                  147       1331         522  
2018-07           309                  122       1378         500  
2018-08           307                  120       1358         522  
2018-09           300                  118       1368         514

In `pandas`, we call a higher-dimensional table as Multi-Index DataFrame. We will cover more on Week 3: Data Wrangling and Visualization course. So, stay tune~

___

# Aggregation Table

## `pd.crosstab()`

Not only for creating a frequency table, `crosstab` can also be used to perform aggregation by adding two more parameters:

- `values`: Values to be aggregated according to the `index` and `columns`
- `aggfunc`: Aggregate function to be used. Most common function: 'count', 'mean', 'median', 'sum', 'max', 'min'


    pd.crosstab(
        index=...,
        columns=...,
        values=...,
        aggfunc=...
    )

<b id='q4'>Business Question 4</b>

Find out the most expensive product `category`, by its mean of `unit_price`.

In other words, let's calculate the mean of `unit_price` of each product `category`.

In [305]:
pd.crosstab(
    index=household['category'], 
    columns='mean_unit_price', 
    values=household['unit_price'],
    aggfunc='mean'
)

col_0                 mean_unit_price
category                             
Fabric Care              17893.793214
Rice                     70013.146313
Sugar/Flavored Syrup     12645.066024

<b id='q5'>Business Question 5</b>

Just like [Business Question 4](#q4), but please also breakdown the mean for each market (`format`)

In [360]:
mean = pd.crosstab(
    index=[household['category'], household['format']],
    columns='mean_unit_price', 
    values=household['unit_price'],
    aggfunc='mean'
)

mean

col_0                             mean_unit_price
category             format                      
Fabric Care          hypermarket     19328.141915
                     minimarket      17757.135774
                     supermarket     17847.557137
Rice                 hypermarket     71205.458458
                     minimarket      67135.569554
                     supermarket     74921.182150
Sugar/Flavored Syrup hypermarket     13539.915728
                     minimarket      12352.135470
                     supermarket     13071.112361

Let's say we want to compare the mean and median side by side:

In [371]:
median = pd.crosstab(
    index=[household['category'], household['format']],
    columns='median_unit_price', 
    values=household['unit_price'],
    aggfunc='median'
)

mean.compare(median, keep_equal=True, keep_shape=True)


# combine = pd.concat([mean.set_index('format'), median.set_index('format')], 
#                    axis='columns', keys=['mean', 'median'])

# combine

ValueError: Can only compare identically-labeled DataFrame objects

📝 Notes on mean and median:

- Both are measures of central tendency
- Mean = average values of data, Median = middle value (50%) of data
- Median is prefered when the data is skewed (contains extremely large values)

<b id='q6'>Business Question 6</b>

Construct an aggregation table to show the sum of `quantity` across each product `category` and `format`. Find out how many percentages of **Rice** is sold in **hypermarket**?

Note: Don't forget we can use parameter `margins` and `normalize` too.

In [393]:
haha = pd.crosstab(
    index=[household['category'], household['format']],
    columns='sum unit', 
    values=household['unit_price'],
    aggfunc='sum',
    margins=True,
    normalize='columns'
)*100

haha.loc[['Rice']] 

col_0                  sum unit        All
category format                           
Rice     hypermarket   3.978835   3.978835
         minimarket   26.616662  26.616662
         supermarket  16.398030  16.398030

## Knowledge Check: Aggregation Table

_Estimated time required: 20 minutes_

1. Create a cross-tab showing the median of `unit_price` across each `sub_category` and `format`. Add a subtotal to both the row and column. Answer the following questions:

    - Overall by median, Sugar is cheapest at ...
    - Overall by median, Detergent is most expensive at ...

In [430]:
# your code here
k1 = pd.crosstab(
    index=[household['sub_category'], household['format']],
    values=household['unit_price'],
    columns='Median',
    aggfunc='median'
)

k1.loc[['Sugar']].sort_values(by='format', ascending=True).head(1)

col_0                      Median
sub_category format              
Sugar        hypermarket  12250.0

2. In which `quarterly` do the hypermarket achieve its highest total sales (sum of `subtotal`)?. **Hint**: Create a new column `subtotal` which calculate the total price of each item in a transaction by multiplying the `unit_price` with the `quantity`.


<!--
# float formatting, round to 3 decimal places
pd.options.display.float_format = '{:.3f}'.format
-->

In [478]:
# your code here
household['quarter'] = household['purchase_time'].dt.to_period('Q')
household['subtotal'] = household['unit_price'] * household['quantity']

k2 = pd.crosstab(
    index=[household['format'], household['quarter']],
    values=household['subtotal'],
    columns='Sum total',
    aggfunc='sum'
).sort_values(by='format', ascending=False)

k2.loc[['hypermarket']].sort_values(by='Sum total', ascending=False).head(1)

col_0                Sum total wkwkkwwk
format      quarter                    
hypermarket 2018Q1          61795966.99

💭 **Bonus challenge:**

Produce a cross-tabulation table where we calculate the sum of `quantity` using:

- `quarterly` and `weekday` as the `index`
- `format` and `category` as the `columns`

Note: this will return a Multi-Index DataFrame.

In [453]:
pd.crosstab(
    index=[household['quarter'], household['weekday']],
    columns=[household['format'], household['category']],
)

format            hypermarket                            minimarket       \
category          Fabric Care Rice Sugar/Flavored Syrup Fabric Care Rice   
quarter weekday                                                            
2017Q4  Monday             59   20                   50         733  233   
        Tuesday            69   20                   54         765  221   
        Wednesday          63   21                   49         843  205   
        Thursday           57   19                   49         754  225   
        Friday            109   32                   75         955  245   
        Saturday          128   42                   76        1041  280   
        Sunday            152   71                  100        1098  355   
2018Q1  Monday             67   28                   55         704  233   
        Tuesday            67   29                   55         772  269   
        Wednesday          63   24                   63         787  239   
        Thursday           63   33                   58         783  226   
        Friday            102   57                   62        1016  231   
        Saturday          170   52                   96         942  272   
        Sunday            158   59                  122         932  235   
2018Q2  Monday             64   27                   53         717  221   
        Tuesday            83   22                   55         789  285   
        Wednesday          69   27                   57         740  243   
        Thursday           82   20                   51         760  246   
        Friday             76   30                   66         942  239   
        Saturday          128   47                   74         993  257   
        Sunday            140   69                   81         970  294   
2018Q3  Monday             74   36                   40         849  271   
        Tuesday            72   23                   51         782  236   
        Wednesday          61   30                   47         856  247   
        Thursday           64   28                   47         730  205   
        Friday            111   35                   47         971  258   
        Saturday          112   34                   57        1024  291   
        Sunday            148   64                   71        1097  326   

format                                 supermarket                            
category          Sugar/Flavored Syrup Fabric Care Rice Sugar/Flavored Syrup  
quarter weekday                                                               
2017Q4  Monday                     510         266  114                  213  
        Tuesday                    545         267  131                  255  
        Wednesday                  522         232  108                  240  
        Thursday                   517         275  127                  238  
        Friday                     530         336  133                  225  
        Saturday                   531         362  187                  276  
        Sunday                     641         436  211                  304  
2018Q1  Monday                     510         244  128                  221  
        Tuesday                    489         286  134                  230  
        Wednesday                  456         274  106                  228  
        Thursday                   529         276  159                  236  
        Friday                     501         390  141                  300  
        Saturday                   575         458  151                  349  
        Sunday                     544         446  194                  321  
2018Q2  Monday                     525         320  111                  240  
        Tuesday                    536         318  139                  232  
        Wednesday                  550         276  124                  200  
        Thursday                   560         280  138        

## Pivot Table

Actually if our data is already a `DataFrame` object, using `pd.pivot_table()` can be more convenient compared to a `pd.crosstab()`.

Parameters:

- `data`: the `DataFrame` object (**not available in `pd.crosstab()`**)
- `index`: the column to be used as rows
- `columns`: the column to be used as columns
- `values`: the values used to fill in the table
- `aggfunc`: the aggregation function (**default: 'mean'**)

Syntax:

```
pd.pivot_table(
    data=...,
    index=...,
    columns=...,
    values=...,
    aggfunc=...
)
```

OR

```
data.pivot_table(
    index=...,
    columns=...,
    values=...,
    aggfunc=...
)
```

Let's compare between `pivot_table` and `crosstab`. Consider the following aggregation table created using `crosstab`:

In [457]:
pd.crosstab(
    index=household['weekday'], 
    columns=[household['format'], household['sub_category']], 
    values=household['unit_price'],
    aggfunc='mean'
)

format         hypermarket                                minimarket  \
sub_category     Detergent          Rice         Sugar     Detergent   
weekday                                                                
Monday        19138.560606  66298.693694  13358.891903  17560.317294   
Tuesday       19389.388316  74470.872340  13794.590698  17901.031532   
Wednesday     19372.105469  68216.509804  13234.430556  17809.980058   
Thursday      19879.209774  68319.880000  13324.351220  17987.282458   
Friday        18891.041457  73596.727273  12885.248000  17694.003432   
Saturday      19970.335818  72381.885714  13939.891089  17747.660396   
Sunday        18831.241756  72182.653992  13897.502674  17649.690262   

format                                     supermarket                \
sub_category          Rice         Sugar     Detergent          Rice   
weekday                                                                
Monday        68546.853862  12457.940470  18133.631542  73603.415629   
Tuesday       68448.321464  12286.595116  17944.182547  73134.498062   
Wednesday     66220.127409  12255.598561  17797.291157  71961.800562   
Thursday      67578.548780  12511.782441  17906.231623  75706.441218   
Friday        65973.959918  12249.178311  17535.585915  77468.788054   
Saturday      66611.120909  12313.932787  17773.888751  74740.144531   
Sunday        66708.621488  12385.304045  17939.162317  76545.073587   

format                      
sub_category         Sugar  
weekday                     
Monday        13024.603859  
Tuesday       13436.082699  
Wednesday     13234.805461  
Thursday      12694.992999  
Friday        12982.341464  
Saturday      13001.971354  
Sunday        13114.061792

Let's re-create the table above with `pd.pivot_table()`.

Note: Default `aggfunc='mean'`

In [455]:
pd.pivot_table(
    data=household,
    index='weekday',
    columns=['format', 'sub_category'],
    values='unit_price'
)

format         hypermarket                                minimarket  \
sub_category     Detergent          Rice         Sugar     Detergent   
weekday                                                                
Monday        19138.560606  66298.693694  13358.891903  17560.317294   
Tuesday       19389.388316  74470.872340  13794.590698  17901.031532   
Wednesday     19372.105469  68216.509804  13234.430556  17809.980058   
Thursday      19879.209774  68319.880000  13324.351220  17987.282458   
Friday        18891.041457  73596.727273  12885.248000  17694.003432   
Saturday      19970.335818  72381.885714  13939.891089  17747.660396   
Sunday        18831.241756  72182.653992  13897.502674  17649.690262   

format                                     supermarket                \
sub_category          Rice         Sugar     Detergent          Rice   
weekday                                                                
Monday        68546.853862  12457.940470  18133.631542  73603.415629   
Tuesday       68448.321464  12286.595116  17944.182547  73134.498062   
Wednesday     66220.127409  12255.598561  17797.291157  71961.800562   
Thursday      67578.548780  12511.782441  17906.231623  75706.441218   
Friday        65973.959918  12249.178311  17535.585915  77468.788054   
Saturday      66611.120909  12313.932787  17773.888751  74740.144531   
Sunday        66708.621488  12385.304045  17939.162317  76545.073587   

format                      
sub_category         Sugar  
weekday                     
Monday        13024.603859  
Tuesday       13436.082699  
Wednesday     13234.805461  
Thursday      12694.992999  
Friday        12982.341464  
Saturday      13001.971354  
Sunday        13114.061792

If we want to summarize the table by `max` of `unit_price`, just specify the `aggfunc` parameter:

In [458]:
pd.pivot_table(
    data=household,
    index='weekday', 
    columns=['format', 'sub_category'], 
    values='unit_price',
    aggfunc='max'
)

format       hypermarket                    minimarket                     \
sub_category   Detergent      Rice    Sugar  Detergent      Rice    Sugar   
weekday                                                                     
Monday           57000.0  175000.0  59950.0    57000.0  219400.0  57000.0   
Tuesday          59500.0  213725.0  59900.0    59000.0  215000.0  50000.0   
Wednesday        50000.0  214000.0  50000.0    55675.0  218592.0  57000.0   
Thursday         54475.0  128000.0  49600.0    57500.0  215000.0  59900.0   
Friday           55440.0  141200.0  43815.0    56000.0  215000.0  47200.0   
Saturday         58470.0  195650.0  59500.0    58000.0  215000.0  56000.0   
Sunday           55560.0  129000.0  59900.0    57100.0  219000.0  58000.0   

format       supermarket                     
sub_category   Detergent      Rice    Sugar  
weekday                                      
Monday           53200.0  219000.0  59500.0  
Tuesday          59900.0  179900.0  59216.0  
Wednesday        53900.0  192000.0  59900.0  
Thursday         55000.0  192000.0  59950.0  
Friday           57200.0  218900.0  58500.0  
Saturday         56400.0  214000.0  59900.0  
Sunday           59600.0  218000.0  59900.0

Let's recreate a frequency table below which shows the number of transactions, using `pd.pivot_table()`:

| category                 |   hypermarket |   minimarket |   supermarket |   Total |
|:-------------------------|--------------:|-------------:|--------------:|--------:|
| **Fabric Care**          |          2611 |        24345 |          9044 |   36000 |
| **Rice**                 |           999 |         7088 |          3913 |   12000 |
| **Sugar/Flavored Syrup** |          1761 |        15370 |          6869 |   24000 |
| **Total**                |          5371 |        46803 |         19826 |   72000 |

- index: ...
- columns: ...
- aggfunc: ...
- values: ...
- "Total" in row and column using ...

In [488]:
pd.pivot_table(
    data=household,
    index='category',
    values='subtotal',
    columns='format',
    margins_name='Total',
    margins=True
)

format                  hypermarket    minimarket   supermarket         Total
category                                                                     
Fabric Care            28412.196856  23337.270487  24738.854488  24057.452666
Rice                  102479.930931  86508.830559  94261.849478  90366.554917
Sugar/Flavored Syrup   23785.120398  20503.070398  22550.114868  21329.772002
Total                  40671.624280  31973.434032  37702.085949  34199.742820

**Note:** parameter `margins` and `margins_name` are available on both `crosstab` and `pivot_table`, but parameter `normalize` only available on `crosstab`

___

# Summary: Tables in `pandas` 

## Frequency Tables

Usage: calculate number of rows for categorical column

Methods:

- For one column, use `.value_counts()` (Series)
- For one or more columns:
    - `pd.crosstab(index, columns)`
    - `pd.pivot_table(index, columns, aggfunc='count')` but rarely used

## Aggregation Tables

Usage: aggregate numerical column which broken down by categorical column

Methods:

- `pd.crosstab(index, columns, values, aggfunc)`
- `pd.pivot_table(data, index, columns, values, aggfunc)`

## `crosstab` vs `pivot_table`

The main differences between `crosstab` and `pivot_table` can be summarized as follows:

|                                                                                    | `pd.crosstab()` | `pd.pivot_table()` |
|------------------------------------------------------------------------------------|-----------------|--------------------|
|                                                                          **Input** | Array of values |          DataFrame |
|                                                              **Default `aggfunc`** |       `'count'` |           `'mean'` |
|                                                          **Parameter `normalize`** |       Available |      Not Available |
| [**Computation Time**](https://ramiro.org/notebook/pandas-crosstab-groupby-pivot/) | Relatively Slower |  Relatively Faster |

In [491]:
# crosstab expect a list/array of values, so the input doesn't have to be a DataFrame
import numpy as np

pd.crosstab(
    index=np.array(['Sugar', 'Sugar', 'Rice', 'Rice']),
    columns=np.array(['hypermarket', 'hypermarket', 'minimarket', 'hypermarket'])
)

col_0  hypermarket  minimarket
row_0                         
Rice             1           1
Sugar            2           0

# Missing Values and Duplicates

During the data exploration and preparation phase, it is likely we come across some problematic details in our data, such as: 

- Value of _-1_ for the _age_ column
- Value of _blank_ for the _customer segment_ column
- Value of _None_ for the _loan duration_ column
- etc

All of these are examples of "untidy" data, which is rather common depending on the data collection and recording process in a company.

## Missing Values

Let's import `household_untidy.csv`, which is a manipulated version of `household` DataFrame. 

**Note**: If you're curious on how to inject the missing values, check out our main materials on Section 3.

In [492]:
household_untidy = pd.read_csv("data_input/household_untidy.csv", index_col='receipts_item_id')
household_untidy['purchase_time'] = household_untidy['purchase_time'].astype('datetime64')
household_untidy.head()

purchase_time category      format  unit_price  \
receipts_item_id                                                        
32000000                         NaT      NaN         NaN         NaN   
32000001                         NaT      NaN         NaN         NaN   
32030785         2018-07-17 18:05:00     Rice  minimarket     63500.0   
32000002                         NaT      NaN         NaN         NaN   
32000003                         NaT      NaN         NaN         NaN   

                  discount  quantity  weekday  Unnamed: 8  
receipts_item_id                                           
32000000               NaN       NaN      NaN         NaN  
32000001               NaN       NaN      NaN         NaN  
32030785               0.0       1.0  Tuesday         NaN  
32000002               NaN       NaN      NaN         NaN  
32000003               NaN       NaN      NaN         NaN

Missing Values:

- `NaN`: Not a Number, for object and category
- `nan`: not a number, for numeric
- `NaT`: Not a Time, for datetime64

### Check Missing Values

A common method to detect missing values:

- `.isna()`: returns `True` if the values are missing.
- `.notna()`: returns `True` if the values are **not** missing.

In [493]:
household_untidy.isna()

purchase_time  category  format  unit_price  discount  \
receipts_item_id                                                          
32000000                   True      True    True        True      True   
32000001                   True      True    True        True      True   
32030785                  False     False   False       False     False   
32000002                   True      True    True        True      True   
32000003                   True      True    True        True      True   
32000004                   True      True    True        True      True   
32369294                  False     False   False       False     False   
31885876                  False     False   False       False     False   
31930241                  False     False   False       False     False   
32418582                  False     False   False       False     False   
32561236                  False     False   False       False     False   
32030785                  False     False   False       False     False   
32935097                  False     False   False       False     False   
32593606                  False     False   False       False     False   
32573843                  False     False   False       False     False   
31913062                  False     False   False       False     False   
31125365                  False     False   False       False     False   
32856560                  False     False   False       False     False   
32552145                  False     False   False       False     False   
32369065                  False     False   False       False     False   

                  quantity  weekday  Unnamed: 8  
receipts_item_id                                 
32000000              True     True        True  
32000001              True     True        True  
32030785             False    False        True  
32000002              True     True        True  
32000003              True     True        True  
32000004              True     True        True  
32369294             False    False        True  
31885876             False     True        True  
31930241             False    False        True  
32418582             False    False        True  
32561236             False    False        True  
32030785             False    False        True  
32935097             False    False        True  
32593606             False    False        True  
32573843             False    False        True  
31913062             False    False        True  
31125365             False    False        True  
32856560             False    False        True  
32552145             False    False        True  
32369065             False    False        True

Count the **number** of missing values across each column:

- `True` will be treated as 1
- `False` will be treated as 0

In [494]:
household_untidy.isna().sum()

purchase_time     5
category          5
format            5
unit_price        5
discount          5
quantity          5
weekday           6
Unnamed: 8       20
dtype: int64

Usually, we count the **percentages** of missing values across each column:

In [496]:
percent_missing = household_untidy.isna().sum() * 100 / len(household_untidy)

missing_value_df = pd.DataFrame({'column_name': household_untidy.columns,
                                 'percent_missing': percent_missing})

missing_value_df

column_name  percent_missing
purchase_time  purchase_time             25.0
category            category             25.0
format                format             25.0
unit_price        unit_price             25.0
discount            discount             25.0
quantity            quantity             25.0
weekday              weekday             30.0
Unnamed: 8        Unnamed: 8            100.0

### Treatment for Missing Value

Three most common ways to deal with missing value:

1. Delete column
2. Delete row
3. Imputation

#### Delete Column

Use `.drop(columns = ...)` method

In [497]:
household_untidy = household_untidy.drop(columns='Unnamed: 8')
household_untidy.head()

purchase_time category      format  unit_price  \
receipts_item_id                                                        
32000000                         NaT      NaN         NaN         NaN   
32000001                         NaT      NaN         NaN         NaN   
32030785         2018-07-17 18:05:00     Rice  minimarket     63500.0   
32000002                         NaT      NaN         NaN         NaN   
32000003                         NaT      NaN         NaN         NaN   

                  discount  quantity  weekday  
receipts_item_id                               
32000000               NaN       NaN      NaN  
32000001               NaN       NaN      NaN  
32030785               0.0       1.0  Tuesday  
32000002               NaN       NaN      NaN  
32000003               NaN       NaN      NaN

⚠️ **Warning**: When removing a column, it is necessary to pay attention to the business case, will the discarded column eliminate or reduce the information of our data?

#### Delete Row

When we are certain that the rows with missing value can be safely dropped, we can use `.dropna()` method.

- `.dropna(how='any')`: drops row if **minimum 1 column** contain missing value

- `.dropna(how='all')`: drops row if **all values** are missing

- `.dropna(thresh=...)`: drops row if non-missing values < `thresh` 

In [498]:
household_untidy.dropna(how='any')#.shape[0]

purchase_time category       format  unit_price  \
receipts_item_id                                                         
32030785         2018-07-17 18:05:00     Rice   minimarket     63500.0   
32369294         2018-07-22 21:19:00     Rice  supermarket    128000.0   
31930241         2018-07-15 12:12:00     Rice  supermarket     64000.0   
32418582         2018-07-24 08:27:00     Rice   minimarket     65000.0   
32561236         2018-07-26 11:28:00     Rice  supermarket    124500.0   
32030785         2018-07-17 18:05:00     Rice   minimarket     63500.0   
32935097         2018-07-29 18:18:00     Rice  supermarket     66500.0   
32593606         2018-07-25 12:48:00     Rice   minimarket     62500.0   
32573843         2018-07-26 16:41:00     Rice   minimarket     62500.0   
31913062         2018-07-14 21:17:00     Rice  supermarket     64000.0   
31125365         2018-07-02 15:39:00     Rice   minimarket    112500.0   
32856560         2018-07-31 05:51:00     Rice  hypermarket     77750.0   
32552145         2018-07-26 11:43:00     Rice   minimarket     66500.0   
32369065         2018-07-23 14:22:00     Rice   minimarket     62900.0   

                  discount  quantity    weekday  
receipts_item_id                                 
32030785               0.0       1.0    Tuesday  
32369294               0.0       1.0     Sunday  
31930241               0.0       3.0     Sunday  
32418582               0.0       1.0    Tuesday  
32561236               0.0       1.0   Thursday  
32030785               0.0       1.0    Tuesday  
32935097               0.0       1.0     Sunday  
32593606               0.0       1.0  Wednesday  
32573843               0.0       1.0   Thursday  
31913062               0.0       3.0   Saturday  
31125365               0.0       1.0     Monday  
32856560               0.0       1.0    Tuesday  
32552145               0.0       1.0   Thursday  
32369065               0.0       1.0     Monday

In [499]:
household_untidy.dropna(how='all')#.shape[0]

purchase_time category       format  unit_price  \
receipts_item_id                                                         
32030785         2018-07-17 18:05:00     Rice   minimarket     63500.0   
32369294         2018-07-22 21:19:00     Rice  supermarket    128000.0   
31885876         2018-07-15 16:17:00     Rice   minimarket    102750.0   
31930241         2018-07-15 12:12:00     Rice  supermarket     64000.0   
32418582         2018-07-24 08:27:00     Rice   minimarket     65000.0   
32561236         2018-07-26 11:28:00     Rice  supermarket    124500.0   
32030785         2018-07-17 18:05:00     Rice   minimarket     63500.0   
32935097         2018-07-29 18:18:00     Rice  supermarket     66500.0   
32593606         2018-07-25 12:48:00     Rice   minimarket     62500.0   
32573843         2018-07-26 16:41:00     Rice   minimarket     62500.0   
31913062         2018-07-14 21:17:00     Rice  supermarket     64000.0   
31125365         2018-07-02 15:39:00     Rice   minimarket    112500.0   
32856560         2018-07-31 05:51:00     Rice  hypermarket     77750.0   
32552145         2018-07-26 11:43:00     Rice   minimarket     66500.0   
32369065         2018-07-23 14:22:00     Rice   minimarket     62900.0   

                  discount  quantity    weekday  
receipts_item_id                                 
32030785               0.0       1.0    Tuesday  
32369294               0.0       1.0     Sunday  
31885876               0.0       1.0        NaN  
31930241               0.0       3.0     Sunday  
32418582               0.0       1.0    Tuesday  
32561236               0.0       1.0   Thursday  
32030785               0.0       1.0    Tuesday  
32935097               0.0       1.0     Sunday  
32593606               0.0       1.0  Wednesday  
32573843               0.0       1.0   Thursday  
31913062               0.0       3.0   Saturday  
31125365               0.0       1.0     Monday  
32856560               0.0       1.0    Tuesday  
32552145               0.0       1.0   Thursday  
32369065               0.0       1.0     Monday

In [500]:
# example: thresh=1, minimal 1 non-NA value present so that it doesn't drop the row
household_untidy.dropna(thresh=1)

purchase_time category       format  unit_price  \
receipts_item_id                                                         
32030785         2018-07-17 18:05:00     Rice   minimarket     63500.0   
32369294         2018-07-22 21:19:00     Rice  supermarket    128000.0   
31885876         2018-07-15 16:17:00     Rice   minimarket    102750.0   
31930241         2018-07-15 12:12:00     Rice  supermarket     64000.0   
32418582         2018-07-24 08:27:00     Rice   minimarket     65000.0   
32561236         2018-07-26 11:28:00     Rice  supermarket    124500.0   
32030785         2018-07-17 18:05:00     Rice   minimarket     63500.0   
32935097         2018-07-29 18:18:00     Rice  supermarket     66500.0   
32593606         2018-07-25 12:48:00     Rice   minimarket     62500.0   
32573843         2018-07-26 16:41:00     Rice   minimarket     62500.0   
31913062         2018-07-14 21:17:00     Rice  supermarket     64000.0   
31125365         2018-07-02 15:39:00     Rice   minimarket    112500.0   
32856560         2018-07-31 05:51:00     Rice  hypermarket     77750.0   
32552145         2018-07-26 11:43:00     Rice   minimarket     66500.0   
32369065         2018-07-23 14:22:00     Rice   minimarket     62900.0   

                  discount  quantity    weekday  
receipts_item_id                                 
32030785               0.0       1.0    Tuesday  
32369294               0.0       1.0     Sunday  
31885876               0.0       1.0        NaN  
31930241               0.0       3.0     Sunday  
32418582               0.0       1.0    Tuesday  
32561236               0.0       1.0   Thursday  
32030785               0.0       1.0    Tuesday  
32935097               0.0       1.0     Sunday  
32593606               0.0       1.0  Wednesday  
32573843               0.0       1.0   Thursday  
31913062               0.0       3.0   Saturday  
31125365               0.0       1.0     Monday  
32856560               0.0       1.0    Tuesday  
32552145               0.0       1.0   Thursday  
32369065               0.0       1.0     Monday

#### Imputation

Fill missing value with some value, using `.fillna()` method

First:

- Impute `category`, `format`, `discount`  with value 'Missing'
- Impute `unit_price` with value 0
- Impute `quantity` with value 0

In [501]:
household_untidy[['category', 'format', 'discount']] = household_untidy[['category', 'format', 'discount']].fillna('Missing')
household_untidy['unit_price'] = household_untidy['unit_price'].fillna(0)
household_untidy['quantity'] = household_untidy['quantity'].replace(np.nan, 0) #.fillna(0)
household_untidy.head()

purchase_time category      format  unit_price  \
receipts_item_id                                                        
32000000                         NaT  Missing     Missing         0.0   
32000001                         NaT  Missing     Missing         0.0   
32030785         2018-07-17 18:05:00     Rice  minimarket     63500.0   
32000002                         NaT  Missing     Missing         0.0   
32000003                         NaT  Missing     Missing         0.0   

                 discount  quantity  weekday  
receipts_item_id                              
32000000          Missing       0.0      NaN  
32000001          Missing       0.0      NaN  
32030785                0       1.0  Tuesday  
32000002          Missing       0.0      NaN  
32000003          Missing       0.0      NaN

Note, these two codes are functionally identical:

- `.fillna(0)`
- `.replace(np.nan, 0)`

Second: impute `purchase_time` with the next value. Use parameter `method='bfill'` to **backward fill** the missing values in the dataset

In [502]:
household_untidy['purchase_time'] = household_untidy['purchase_time'].fillna(method='bfill')
household_untidy.head(7)

purchase_time category       format  unit_price  \
receipts_item_id                                                         
32000000         2018-07-17 18:05:00  Missing      Missing         0.0   
32000001         2018-07-17 18:05:00  Missing      Missing         0.0   
32030785         2018-07-17 18:05:00     Rice   minimarket     63500.0   
32000002         2018-07-22 21:19:00  Missing      Missing         0.0   
32000003         2018-07-22 21:19:00  Missing      Missing         0.0   
32000004         2018-07-22 21:19:00  Missing      Missing         0.0   
32369294         2018-07-22 21:19:00     Rice  supermarket    128000.0   

                 discount  quantity  weekday  
receipts_item_id                              
32000000          Missing       0.0      NaN  
32000001          Missing       0.0      NaN  
32030785                0       1.0  Tuesday  
32000002          Missing       0.0      NaN  
32000003          Missing       0.0      NaN  
32000004          Missing       0.0      NaN  
32369294                0       1.0   Sunday

💡 **Tips on Imputation**:

For categorical column, do either:
- Make NA as one category
- Fill using a central value (mode as the most frequent category)
- Fill using predictive/machine learning model (Classification problem)

For numerical column, do either:
- Fill using a central value (mean or median)
- Fill using predictive/machine learning model (Regression problem)

Case: In a dataframe where `salary` is missing but the bank has data about the customer's occupation / profession, years of experience, years of education, seniority level, age, and industry, then a machine learning model can offer a viable alternative to the mean imputation approach.

## Duplicated Data

### Check Duplicates

Use method `.duplicated()` and combine with subsetting operation.

Parameter `keep`: 

- `keep='first'` (default): Mark duplicates as True except for the **first occurrence**.
- `keep='last'`: Mark duplicates as True except for the **last occurrence**.
- `keep=False`: Mark all duplicates as True.

In [503]:
cond = household_untidy.duplicated(keep=False)
household_untidy[cond]

purchase_time category      format  unit_price  \
receipts_item_id                                                        
32000000         2018-07-17 18:05:00  Missing     Missing         0.0   
32000001         2018-07-17 18:05:00  Missing     Missing         0.0   
32030785         2018-07-17 18:05:00     Rice  minimarket     63500.0   
32000002         2018-07-22 21:19:00  Missing     Missing         0.0   
32000003         2018-07-22 21:19:00  Missing     Missing         0.0   
32000004         2018-07-22 21:19:00  Missing     Missing         0.0   
32030785         2018-07-17 18:05:00     Rice  minimarket     63500.0   

                 discount  quantity  weekday  
receipts_item_id                              
32000000          Missing       0.0      NaN  
32000001          Missing       0.0      NaN  
32030785                0       1.0  Tuesday  
32000002          Missing       0.0      NaN  
32000003          Missing       0.0      NaN  
32000004          Missing       0.0      NaN  
32030785                0       1.0  Tuesday

### Treatment for Duplicated Data

Drop the duplicated row by using `.drop_duplicates()` method.

Parameter `keep`: 

- `keep='first'` (default): Drop duplicates except for the **first occurrence**.
- `keep='last'`: Drop duplicates except for the **last occurrence**.
- `keep=False`: Drop all duplicates.

In [504]:
household_untidy.drop_duplicates(keep='last')

purchase_time category       format  unit_price  \
receipts_item_id                                                         
32000001         2018-07-17 18:05:00  Missing      Missing         0.0   
32000004         2018-07-22 21:19:00  Missing      Missing         0.0   
32369294         2018-07-22 21:19:00     Rice  supermarket    128000.0   
31885876         2018-07-15 16:17:00     Rice   minimarket    102750.0   
31930241         2018-07-15 12:12:00     Rice  supermarket     64000.0   
32418582         2018-07-24 08:27:00     Rice   minimarket     65000.0   
32561236         2018-07-26 11:28:00     Rice  supermarket    124500.0   
32030785         2018-07-17 18:05:00     Rice   minimarket     63500.0   
32935097         2018-07-29 18:18:00     Rice  supermarket     66500.0   
32593606         2018-07-25 12:48:00     Rice   minimarket     62500.0   
32573843         2018-07-26 16:41:00     Rice   minimarket     62500.0   
31913062         2018-07-14 21:17:00     Rice  supermarket     64000.0   
31125365         2018-07-02 15:39:00     Rice   minimarket    112500.0   
32856560         2018-07-31 05:51:00     Rice  hypermarket     77750.0   
32552145         2018-07-26 11:43:00     Rice   minimarket     66500.0   
32369065         2018-07-23 14:22:00     Rice   minimarket     62900.0   

                 discount  quantity    weekday  
receipts_item_id                                
32000001          Missing       0.0        NaN  
32000004          Missing       0.0        NaN  
32369294                0       1.0     Sunday  
31885876                0       1.0        NaN  
31930241                0       3.0     Sunday  
32418582                0       1.0    Tuesday  
32561236                0       1.0   Thursday  
32030785                0       1.0    Tuesday  
32935097                0       1.0     Sunday  
32593606                0       1.0  Wednesday  
32573843                0       1.0   Thursday  
31913062                0       3.0   Saturday  
31125365                0       1.0     Monday  
32856560                0       1.0    Tuesday  
32552145                0       1.0   Thursday  
32369065                0       1.0     Monday

Suppose we want to show unique observation of `receipt_id`:

In [505]:
household[household['receipt_id'].duplicated()]

receipt_id  receipts_item_id       purchase_time              category  \
353       9270720          31330334 2018-07-05 13:13:00                  Rice   
443       9622257          32369293 2018-07-22 21:19:00                  Rice   
634       9655239          32451761 2018-07-21 18:04:00                  Rice   
682       9367922          31616632 2018-07-10 22:56:00                  Rice   
996       9623282          32371114 2018-07-21 19:18:00                  Rice   
...           ...               ...                 ...                   ...   
71913     5955640          18149296 2017-12-31 09:28:00  Sugar/Flavored Syrup   
71935     5737076          17437723 2017-12-13 14:13:00  Sugar/Flavored Syrup   
71940     5865653          17834851 2017-12-24 10:47:00  Sugar/Flavored Syrup   
71954     5805533          17651393 2017-12-19 13:07:00  Sugar/Flavored Syrup   
71965     5935719          18295911 2017-12-28 14:23:00  Sugar/Flavored Syrup   

      sub_category       format  unit_price  discount  quantity yearmonth  \
353           Rice  supermarket     64000.0         0         1   2018-07   
443           Rice  supermarket    159900.0         0         1   2018-07   
634           Rice  supermarket     59900.0      4100         1   2018-07   
682           Rice   minimarket     64000.0         0         1   2018-07   
996           Rice   minimarket     28500.0         0         2   2018-07   
...            ...          ...         ...       ...       ...       ...   
71913        Sugar   minimarket     11500.0         0         2   2017-12   
71935        Sugar   minimarket     12200.0       600         2   2017-12   
71940        Sugar  supermarket     12990.0         0         1   2017-12   
71954        Sugar   minimarket     12500.0         0         1   2017-12   
71965        Sugar   minimarket     12500.0         0        10   2017-12   

         weekday yearmonth_recreate quarter  subtotal  
353     Thursday            2018-07  2018Q3   64000.0  
443       Sunday            2018-07  2018Q3  159900.0  
634     Saturday            2018-07  2018Q3   59900.0  
682      Tuesday            2018-07  2018Q3   64000.0  
996     Saturday            2018-07  2018Q3   57000.0  
...          ...                ...     ...       ...  
71913     Sunday            2017-12  2017Q4   23000.0  
71935  Wednesday            2017-12  2017Q4   24400.0  
71940     Sunday            2017-12  2017Q4   12990.0  
71954    Tuesday            2017-12  2017Q4   12500.0  
71965   Thursday            2017-12  2017Q4  125000.0  

[2224 rows x 14 columns]

⚠️ **Warning**: Duplicates may mean a different thing from a data point-of-view and a business analyst's point-of-view. You want to be extra careful about whether the duplicates is an intended characteristic of your data, or whether it poses a violation to the business logic.

## Knowledge Check: Missing Values and Duplicates

_Estimated time required: 10 minutes_

Should we drop the duplicates for each of the following case, and why:

1. A medical center collects anonymized heart rate monitoring data from patients. It has duplicate observations collected across a span of 3 months
2. An insurance company uses machine learning to deliver dynamic pricing to its customers. Each row contains the customer's name, occupation / profession and historical health data. It has duplicate observations collected across a span of 3 months
3. On our original `household` data, check for duplicate observations. Would you have drop the duplicated rows?

In [510]:
# illustration data for question 1
heart = pd.DataFrame({
    'rate': [70, 80, 90, 75, 95, 70, 85, 90]
})

heart.drop_duplicates(keep='first')

rate
0    70
1    80
2    90
3    75
4    95
6    85

In [557]:
# illustration data for question 2
insurance = pd.DataFrame({
    'cust_id': ['C1', 'C2', 'C3', 'C4', 'C1'],
    'occupation': ['Employee', 'Employee', 'Student', 'Student', 'Employee'],
    'health': ['Good', 'Ok', 'Good', 'Ok', 'Ok']
})

insurance['cust_id'] = insurance['cust_id'].drop_duplicates(keep='first')

insurance.dropna(how='any')


# cond = insurance['cust_id'].duplicated()

# insurance = insurance.drop(columns=insurance[cond])

# insurance[cond]

cust_id occupation health
0      C1   Employee   Good
1      C2   Employee     Ok
2      C3    Student   Good
3      C4    Student     Ok

In [525]:
# reference data for question 3
household = pd.read_csv('data_input/household.csv', index_col='receipts_item_id')
household[household.duplicated(keep=False)]

receipt_id     purchase_time              category  \
receipts_item_id                                                       
31616772             9367922   7/10/2018 22:56                  Rice   
31616632             9367922   7/10/2018 22:56                  Rice   
35486879            10626238   8/31/2018 20:21                  Rice   
35486825            10626238   8/31/2018 20:21                  Rice   
37704383            11250231   9/25/2018 14:15                  Rice   
...                      ...               ...                   ...   
17911936             5890361   12/26/2017 9:50  Sugar/Flavored Syrup   
17110567             5636761   12/4/2017 20:06  Sugar/Flavored Syrup   
17685098             5817089  12/19/2017 22:03  Sugar/Flavored Syrup   
18149296             5955640   12/31/2017 9:28  Sugar/Flavored Syrup   
17651393             5805533  12/19/2017 13:07  Sugar/Flavored Syrup   

                 sub_category       format  unit_price  discount  quantity  \
receipts_item_id                                                             
31616772                 Rice   minimarket     64000.0         0         1   
31616632                 Rice   minimarket     64000.0         0         1   
35486879                 Rice   minimarket     62500.0         0         1   
35486825                 Rice   minimarket     62500.0         0         1   
37704383                 Rice  supermarket     64000.0         0         1   
...                       ...          ...         ...       ...       ...   
17911936                Sugar  supermarket      6250.0         0         1   
17110567                Sugar   minimarket     12500.0         0         1   
17685098                Sugar  hypermarket     11990.0       260         1   
18149296                Sugar   minimarket     11500.0         0         2   
17651393                Sugar   minimarket     12500.0         0         1   

                 yearmonth  
receipts_item_id            
31616772           2018-07  
31616632           2018-07  
35486879           2018-08  
35486825           2018-08  
37704383           2018-09  
...                    ...  
17911936           2017-12  
17110567           2017-12  
17685098           2017-12  
18149296           2017-12  
17651393           2017-12  

[650 rows x 9 columns]

**Answer**:

1. Should (drop/not drop) the duplicates, because ...

2. Should (drop/not drop) the duplicates, because ...

3. Should (drop/not drop) the duplicates, because ...